### PR TITLE
Allow mounts to run in parallel for non-attachable volumes

### PR DESCRIPTION
**Pull request by saad-ali**
This PR:
- Fixes https://github.com/kubernetes/kubernetes/issues/28616
  - Enables mount volume operations to run in parallel for non-attachable volume plugins.
  - Enables unmount volume operations to run in parallel for all volume plugins.

### CODE REVIEW

File: pkg/volume/util/operationexecutor/operation_executor.go [hunk 1]
```diff
@@ -390,7 +401,7 @@ func (oe *operationExecutor) UnmountDevice(
 	}
 
 	return oe.pendingOperations.Run(
-		string(deviceToDetach.VolumeName), unmountDeviceFunc)
+		string(deviceToDetach.VolumeName), "" /* operationSubName */, unmountDeviceFunc)
```
Comment (dims) [inline pkg/volume/util/operationexecutor/operation_executor.go]: @saad-ali : (trying to learn). So the empty subname "" means that all the UnmountDevice operations will run one after another right?

File: pkg/volume/util/operationexecutor/operation_executor.go [hunk 2]
```diff
@@ -404,7 +415,7 @@ func (oe *operationExecutor) VerifyControllerAttachedVolume(
 	}
 
 	return oe.pendingOperations.Run(
-		string(volumeToMount.VolumeName), verifyControllerAttachedVolumeFunc)
+		string(volumeToMount.VolumeName), "" /* operationSubName */, verifyControllerAttachedVolumeFunc)
```
Comment (dims) [inline pkg/volume/util/operationexecutor/operation_executor.go]: Same question here

File: pkg/util/goroutinemap/goroutinemap.go [hunk 3]
```diff
@@ -43,34 +43,38 @@ const (
...
 	g.cond = sync.NewCond(g)
 	return g
 }
 
 type goRoutineMap struct {
-	operations                map[string]operation
+	operations                map[string]map[string]operation
 	exponentialBackOffOnError bool
 	cond                      *sync.Cond
 	sync.Mutex
```
Comment (smarterclayton) [inline pkg/util/goroutinemap/goroutinemap.go]: nit: we prefer `lock sync.Mutex` because it prevents Mutex from being public to the goRoutineMap, unless you have a need for that.

Comment (smarterclayton) [inline pkg/util/goroutinemap/goroutinemap.go]: I would prefer getting access to Lock being impossible from outside this object without a good justification.

File: pkg/volume/util/operationexecutor/operation_executor.go (hunk 1 above)
Comment (saad-ali) [inline pkg/volume/util/operationexecutor/operation_executor.go]: Correct.

File: pkg/volume/util/operationexecutor/operation_executor.go (hunk 2 above)
Comment (saad-ali) [inline pkg/volume/util/operationexecutor/operation_executor.go]: Yep.

File: pkg/util/goroutinemap/goroutinemap.go (hunk 3 above)
Comment (saad-ali) [inline pkg/util/goroutinemap/goroutinemap.go]: Sure, I'll make the change.

Comment (saad-ali) [inline pkg/util/goroutinemap/goroutinemap.go]: Done.

**Comment by saad-ali** [comment]
PTAL

File: pkg/util/goroutinemap/goroutinemap.go [hunk 4]
```diff
@@ -43,37 +43,42 @@ const (
...
+	// and spawns a new go routine to execute operationFunc.
+	// If an operation with the same operationName and no sub name already
+	// exists, an AlreadyExists or ExponentialBackoff error is returned.
+	// If an operation with the same operationName and operationSubName already
+	// exists, an AlreadyExists or ExponentialBackoff error is returned.
+	// Once the operation is complete, the go routine is terminated and the
+	// operationName/operationSubName is removed from the list of executing
+	// operations allowing a new operation to be started with the same name
+	// without error.
+	Run(operationName, operationSubName string, operationFunc func() error) error
```
Comment (matchstick) [inline pkg/util/goroutinemap/goroutinemap.go]: @saad-ali @smarterclayton The run interface was a nice and simple interface. We are burdending the compelxity of the "sub" logic into callers who don't care. We should add a new routine for the subLogic that allows for running on two dimensions as opposed to one. That also means callers don't have to use a "" second argument that is just confusing.

How about 
```
Run(operationName, operationSubName string, operationFunc func() error) error
RunSubTargets(target string,  targetSub string, operationFunc func() error) error
```

File: pkg/util/goroutinemap/goroutinemap.go [hunk 5]
```diff
@@ -43,37 +43,42 @@ const (
...
-	// Run adds operationName to the list of running operations and spawns a new
-	// go routine to execute the operation. If an operation with the same name
-	// already exists, an error is returned. Once the operation is complete, the
-	// go routine is terminated and the operationName is removed from the list
-	// of executing operations allowing a new operation to be started with the
-	// same name without error.
-	Run(operationName string, operationFunc func() error) error
+// GoroutineMap defines the supported set of operations.
+type GoroutineMap interface {
+	// Run adds operationName/operationSubName to the list of running operations
```
Comment (matchstick) [inline pkg/util/goroutinemap/goroutinemap.go]: This comment  is good at explaining the behaviour but not the motivation. As commented below we should be able to split the routine into two. Then add a comment just for Sub along the lines of: 

```
/// Here we have to build a sub dependendancy. We want to run the routine only when the concatenation of the target and subTarget are both not being operated on by anything.

File: pkg/util/goroutinemap/goroutinemap.go (hunk 4 above)
Comment (smarterclayton) [inline pkg/util/goroutinemap/goroutinemap.go]: Minor nit here - we usually call function variables `fooFn` or `fn` and function type aliases `FooFunc`.

Comment (smarterclayton) [inline pkg/util/goroutinemap/goroutinemap.go]: Agree on splitting.  Alternatively, you could reorder the function like:

```
Run(fn func() error, targets ...string)
```

and have identical targets block (you might have to join the targets to do that or create a special map key, so it's not hugely important to me).  Either way making clear that target has to be unique.

**Comment by coufon** [comment]
The latency of volume gets evidently improved after this fixup. The new result of re-running the test in #28616 is now (create 30 pods):



The latency of volume manager handling 30 pods are reduced down to ~6s (it is ~24s at the beginning)

**Comment by matchstick** [comment]
@saad-ali @thockin @smarterclayton I spent some time reviewing this today and I talked to saad. I think the high level goals of this CL are spot on. 

But the modifications to the goRoutineMap seem a bit too intrusive. We are overloading an elegant datastructure with new functionality that doesn't seem generic. The system as it stands is simple and generic and serves a good purprose. It looks like we are grafting on complexity that could be better served by simply writing a storage specific wrapper around a slice. Then we can have the code be more clear and maintainable. 

saad and I ran some experiments to make sure that the slice will be able to perform well with a 1000000 operations and it looks like it can do it in about 1ms so pursuing this alternative should yield a simpler and less intrusive solution.

**Comment by saad-ali** [comment]
PTAL 
Split GoroutineMap into separate files based on @matchstick's feedback.

File: pkg/volume/util/goroutinemap/goroutinemap.go [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+	}
+
+	// Delete index without preserving order
+	grm.operations[opIndex] = grm.operations[len(grm.operations)-1]
+	grm.operations = grm.operations[:len(grm.operations)-1]
+}
+
+func (grm *goRoutineMap) operationComplete(
+	volumeName api.UniqueVolumeName, podName types.UniquePodName, err *error) {
+	defer grm.cond.Signal()
```
Comment (smarterclayton) [inline pkg/volume/util/goroutinemap/goroutinemap.go]: My brain twisted a bit at this - just to verify, we're going to take and hold lock, but cond will be signaled inside the lock?  So lock dominates cond here as well as in wait?

File: pkg/util/goroutinemap/common/errors.go [hunk 7]
```diff
@@ -0,0 +1,96 @@
+/*
```
Comment (matchstick) [inline pkg/util/goroutinemap/common/errors.go]: I don't think you need this file. I would just move all the exponential errors into the exponential file and then just copy the already exists functionality.

File: pkg/volume/util/goroutinemap/goroutinemap.go [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+		if existingOp.podName != emptyUniquePodName &&
+			podName != emptyUniquePodName &&
+			existingOp.podName != podName {
+			// No match, keep searching
+			continue
+		}
+
+		// Match
+		opExists = true
+		previousOp = existingOp
```
Comment (matchstick) [inline pkg/volume/util/goroutinemap/goroutinemap.go]: If you scope the existingOp out of the loop scoping you can just break.

File: pkg/volume/util/goroutinemap/goroutinemap.go [hunk 9]
```diff
@@ -0,0 +1,252 @@
+/*
```
Comment (matchstick) [inline pkg/volume/util/goroutinemap/goroutinemap.go]: Call this something besides "goroutinemap" It is very specific. Like maybe "AttachMountDependencyTracker" Or something more pithy.

File: pkg/volume/util/goroutinemap/goroutinemap.go [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+
+		if existingOp.podName != emptyUniquePodName &&
+			podName != emptyUniquePodName &&
+			existingOp.podName != podName {
+			// No match, keep searching
+			continue
+		}
+
+		// Match
+		opExists = true
```
Comment (matchstick) [inline pkg/volume/util/goroutinemap/goroutinemap.go]: Don't we mean it doesn't exist yet? Is exists the right word?

Comment (matchstick) [inline pkg/volume/util/goroutinemap/goroutinemap.go]: Oh wait this makes more sense now.

File: pkg/util/goroutinemap/goroutinemap.go [hunk 11]
```diff
@@ -22,19 +22,18 @@ with the same name exists.
...
 	"time"
 
 	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/util/goroutinemap/common"
 	k8sRuntime "k8s.io/kubernetes/pkg/util/runtime"
 )
 
 const (
 	// initialDurationBeforeRetry is the amount of time after an error occurs
-	// that GoRoutineMap will refuse to allow another operation to start with
```
Comment (matchstick) [inline pkg/util/goroutinemap/goroutinemap.go]: @saad-ali saad-ali Maybe do the renaming in a separate PR. IT confuses the changes you make here. And really should not go in a 1.3 fix.

File: pkg/util/goroutinemap/common/exponential_backoff.go [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+// that retries are not permitted.
+type ExponentialBackoff struct {
+	lastError           error
+	lastErrorTime       time.Time
+	durationBeforeRetry time.Duration
+}
+
+// SafeToRetry returns true if the durationBeforeRetry period for the given
+// lastErrorTime has expired.
+func (expBackoff *ExponentialBackoff) SafeToRetry() bool {
```
Comment (matchstick) [inline pkg/util/goroutinemap/common/exponential_backoff.go]: @saad-ali If you had this accept a string you coudl then have it return an error. Then the interface would be extra cool:

err := SafeToRetry(name) 
if err != nil {
    .....
}

File: pkg/util/goroutinemap/goroutinemap.go [hunk 13]
```diff
@@ -59,68 +59,71 @@ type GoRoutineMap interface {
...
+	g.cond = sync.NewCond(g.lock)
 	return g
 }
 
 type goRoutineMap struct {
 	operations                map[string]operation
 	exponentialBackOffOnError bool
 	cond                      *sync.Cond
-	sync.Mutex
+	lock                      *sync.Mutex
```
Comment (matchstick) [inline pkg/util/goroutinemap/goroutinemap.go]: @saad-ali Why are we changing the locking?

File: pkg/util/goroutinemap/common/exponential_backoff.go [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+	if time.Since(expBackoff.lastErrorTime) <= expBackoff.durationBeforeRetry {
+		return false
+	}
+
+	return true
+}
+
+func (expBackoff *ExponentialBackoff) UpdateExponentialBackoffTime(err *error) {
```
Comment (matchstick) [inline pkg/util/goroutinemap/common/exponential_backoff.go]: @saad-ali The func takes in a "ExponentialBackOff struct already" You can just name this "Update"

File: pkg/util/goroutinemap/common/exponential_backoff.go [hunk 15]
```diff
@@ -0,0 +1,77 @@
+/*
```
Comment (matchstick) [inline pkg/util/goroutinemap/common/exponential_backoff.go]: @saad-ali This is a neat piece of code. I would just move it to pkt/util and out of pkg/util/goroutinemap. Share it with the world of kubernetes.

Comment (smarterclayton) [inline pkg/util/goroutinemap/common/exponential_backoff.go]: When you do that consider merging it with `pkg/util/wait/ExponentialBackoff`

File: pkg/controller/volume/attachdetach/reconciler/reconciler.go [hunk 16]
```diff
@@ -114,9 +114,9 @@ func (rc *reconciler) reconciliationLoopFunc() func() {
 						glog.Infof("Started DetachVolume for volume %q from node %q", attachedVolume.VolumeName, attachedVolume.NodeName)
 					}
 					if err != nil &&
-						!goroutinemap.IsAlreadyExists(err) &&
-						!goroutinemap.IsExponentialBackoff(err) {
-						// Ignore goroutinemap.IsAlreadyExists && goroutinemap.IsExponentialBackoff errors, they are expected.
+						!common.IsAlreadyExists(err) &&
```
Comment (matchstick) [inline pkg/controller/volume/attachdetach/reconciler/reconciler.go]: @saad-ali @smarterclayton Even if the implementation is in common (which I am not sure makes sense) I think we can wrap it up in the goRoutine func

File: pkg/controller/volume/attachdetach/reconciler/reconciler.go [hunk 17]
```diff
@@ -114,9 +114,9 @@ func (rc *reconciler) reconciliationLoopFunc() func() {
...
+						!common.IsExponentialBackoff(err) {
```
Comment (matchstick) [inline pkg/controller/volume/attachdetach/reconciler/reconciler.go]: @saad-ali @smarterclayton This should also be in the exponentialbackoff package where I think this code ends up. 
I think the gorutine/common is not needed. We can make it simpler.

File: pkg/controller/volume/persistentvolume/controller.go [hunk 18]
```diff
@@ -129,7 +130,7 @@ type PersistentVolumeController struct {
 	claims  cache.Store
 
 	// Map of scheduled/running operations.
-	runningOperations goroutinemap.GoRoutineMap
+	runningOperations goroutinemap.GoroutineMap
```
Comment (matchstick) [inline pkg/controller/volume/persistentvolume/controller.go]: @saad-ali @justinclayton Let's do the renaming in a separate PR. For 1.3 it doesn't make sense. And it makes the review here much more complicated to see the minimal changes to goRoutineMap.

File: pkg/util/goroutinemap/common/errors.go [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+limitations under the License.
+*/
+
+// Package common contains helper methods and common functionality used by
+// all RoroutineMap implementations.
+package common
+
+import "fmt"
+
+// NewAlreadyExistsError returns a new instance of AlreadyExists error.
```
Comment (matchstick) [inline pkg/util/goroutinemap/common/errors.go]: @saad-ali @smarterclayton I think we can take lines 23-54 (everything but the exponential back off stuff) and just keep it in gorroutinemap pkgand then duplicate for the new datastructure. I know the idea of duplicating this code breaks your heat. it's boiler plate not a big deal to duplicate here.

File: pkg/util/goroutinemap/common/errors.go [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+var _ error = alreadyExistsError{}
+
+func (err alreadyExistsError) Error() string {
+	return fmt.Sprintf(
+		"Failed to create operation with name %q. An operation with that name is already executing.",
+		err.operationName)
+}
+
+// NewExponentialBackoffError returns a new instance of ExponentialBackoff error.
+func NewExponentialBackoffError(
```
Comment (matchstick) [inline pkg/util/goroutinemap/common/errors.go]: @saad-ali @smarterclayton This code is nicely encapsulated. Let's keep it closer to the code that you wrote for the functionality.

File: pkg/util/goroutinemap/common/errors.go [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+	}
+}
+
+// IsExponentialBackoff returns true if an error returned from GoroutineMap
+// indicates that a new operation can not be started because
+// exponentialBackOffOnError is enabled and a previous operation with the same
+// operation failed within the durationBeforeRetry period.
+func IsExponentialBackoff(err error) bool {
+	switch err.(type) {
+	case exponentialBackoffError:
```
Comment (matchstick) [inline pkg/util/goroutinemap/common/errors.go]: @saad-ali @smarterclayton 
why not:

```
return err.(type) == exponentialBackoffErr
```

File: pkg/util/goroutinemap/common/exponential_backoff.go (hunk 15 above)
Comment (matchstick) [inline pkg/util/goroutinemap/common/exponential_backoff.go]: @smarterclayton

File: pkg/util/goroutinemap/common/exponential_backoff.go (hunk 12 above)
Comment (matchstick) [inline pkg/util/goroutinemap/common/exponential_backoff.go]: @smarterclayton

File: pkg/util/goroutinemap/common/util.go [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+// Package common contains helper methods and common functionality used by
+// all RoroutineMap implementations.
+package common
+
+import (
+	"fmt"
+	"runtime"
+)
+
+func RecoverFromPanic(operationName string, err *error) {
```
Comment (matchstick) [inline pkg/util/goroutinemap/common/util.go]: @saad-ali @smarterclayton This code RecoverFromPanic is super generic and belongs somewhere other than goroutinemap pkg

File: pkg/util/goroutinemap/goroutinemap.go (hunk 13 above)
Comment (matchstick) [inline pkg/util/goroutinemap/goroutinemap.go]: @smarterclayton

File: pkg/volume/util/goroutinemap/goroutinemap.go (hunk 9 above)
Comment (matchstick) [inline pkg/volume/util/goroutinemap/goroutinemap.go]: @saad-ali @smarterclayton also we should move it out of goroutinemap directory.

File: pkg/util/goroutinemap/common/errors.go (hunk 20 above)
Comment (matchstick) [inline pkg/util/goroutinemap/common/errors.go]: Here this code means the exponential back off errors.

File: pkg/volume/util/goroutinemap/goroutinemap.go [hunk 23]
```diff
@@ -0,0 +1,252 @@
...
+	}()
+
+	return nil
+}
+
+func (grm *goRoutineMap) getOperation(
+	volumeName api.UniqueVolumeName,
+	podName types.UniquePodName) (uint, error) {
+
+	for i, op := range grm.operations {
```
Comment (matchstick) [inline pkg/volume/util/goroutinemap/goroutinemap.go]: @saad-ali @smarterclayton Do we need a lock here?

File: pkg/volume/util/goroutinemap/goroutinemap.go [hunk 24]
```diff
@@ -0,0 +1,252 @@
...
+			return uint(i), nil
+		}
+	}
+
+	logOperationName := getOperationName(volumeName, podName)
+	return 0, fmt.Errorf("Operation %q not found.", logOperationName)
+}
+
+func (grm *goRoutineMap) deleteOperation(
+	volumeName api.UniqueVolumeName,
```
Comment (matchstick) [inline pkg/volume/util/goroutinemap/goroutinemap.go]: @saad-ali @smarterclayton Do we need a lock here?

File: pkg/volume/util/goroutinemap/goroutinemap.go (hunk 23 above)
Comment (matchstick) [inline pkg/volume/util/goroutinemap/goroutinemap.go]: @saad-ali @smarterclayton I seen now we should comment that we assume the lock is held when we do this.

File: pkg/volume/util/goroutinemap/goroutinemap.go (hunk 24 above)
Comment (matchstick) [inline pkg/volume/util/goroutinemap/goroutinemap.go]: @saad-ali @smarterclayton I seen now we should comment that we assume the lock is held when we do this.

File: pkg/volume/util/goroutinemap/goroutinemap.go (hunk 6 above)
Comment (matchstick) [inline pkg/volume/util/goroutinemap/goroutinemap.go]: @smarterclayton @saad-ali SO defer order according to https://blog.golang.org/defer-panic-and-recover is Last-In is First-Out. Which means that what we are doing here as I under stand it is

1) lock the structure
2) Do all the work in the datastructure required.
3) unlock structure
4) Wakeup anything potentially waiting before this routine takes the lock, or while the lock is taken here. There is a window between the defers where we release the lock and then do the signal but in that case the Wait routine should perform the correct action as the length of the slice should be updated.

Saad is that correct? I recommend commenting the desired order of the defers and the reasoning behind them.

File: pkg/util/goroutinemap/common/util.go (hunk 22 above)
Comment (smarterclayton) [inline pkg/util/goroutinemap/common/util.go]: Arguably `pkg/util/runtime` should have a way to pass a custom message (which is sounds like you are doing here) to a HandleCrash

File: pkg/volume/util/goroutinemap/goroutinemap.go (hunk 6 above)
Comment (saad-ali) [inline pkg/volume/util/goroutinemap/goroutinemap.go]: > My brain twisted a bit at this - just to verify, we're going to take and hold lock, but cond will be signaled inside the lock? So lock dominates cond here as well as in wait?

Cond will be signaled outside the lock here. But the wait is protected by the lock.

So pretty much what @matchstick said, the last operation to occur is signaling any waiting routines.

The wait code (introduced in https://github.com/kubernetes/kubernetes/pull/25399 and https://github.com/kubernetes/kubernetes/pull/28153) is all existing code, so I didn't give it much throught in this PR. Looking at it again, I think it's ok as is. Will add comments to explain the order of execution.

Comment (smarterclayton) [inline pkg/volume/util/goroutinemap/goroutinemap.go]: Just wanted to be sure, thanks.

File: pkg/volume/util/goroutinemap/goroutinemap.go (hunk 8 above)
Comment (saad-ali) [inline pkg/volume/util/goroutinemap/goroutinemap.go]: Done.

File: pkg/util/goroutinemap/goroutinemap.go (hunk 13 above)
Comment (saad-ali) [inline pkg/util/goroutinemap/goroutinemap.go]: To address clayton's earlier comment:

> nit: we prefer lock sync.Mutex because it prevents Mutex from being public to the goRoutineMap, unless you have a need for that.
I would prefer getting access to Lock being impossible from outside this object without a good justification.

File: pkg/util/goroutinemap/common/exponential_backoff.go (hunk 14 above)
Comment (saad-ali) [inline pkg/util/goroutinemap/common/exponential_backoff.go]: Done

File: pkg/util/goroutinemap/goroutinemap.go (hunk 13 above)
Comment (matchstick) [inline pkg/util/goroutinemap/goroutinemap.go]: +1

File: pkg/util/goroutinemap/common/exponential_backoff.go (hunk 15 above)
Comment (saad-ali) [inline pkg/util/goroutinemap/common/exponential_backoff.go]: This logic is pretty specific to the way that goroutinemap operates. The package clayton mentioned, `pkg/util/wait/ExponentialBackoff`, is a more generic mechanism for doing exponential backoff and is already in `pkg/util`. That package takes responsibility for execution (repeatedly execute function with exp backoff until it succeeds). Whereas goroutinemap expects the caller to control function execution and simply blocks it if it is too frequent. Combining the two is a non-trivial task and outside the scope of this PR. And because this backoff mechanism is unique to the way that goroutinemap operates I'd prefer to leave it here until a more through refactor can be done combining the two.

Comment (smarterclayton) [inline pkg/util/goroutinemap/common/exponential_backoff.go]: Thats fine, just wanted to make sure you were aware.

File: pkg/util/goroutinemap/common/errors.go (hunk 21 above)
Comment (saad-ali) [inline pkg/util/goroutinemap/common/errors.go]: It's illegal. `someinterface.(type)` is only permitted in [type switches](https://golang.org/ref/spec#Switch_statements).

We could do type assertion instead:
```GO
func IsExponentialBackoff(err error) bool {
	if err == nil {
		return false
	}
	_, ok := err.(*exponentialBackoffErr)
	return ok
}
```
?

File: pkg/volume/util/goroutinemap/goroutinemap.go (hunk 23 above)
Comment (saad-ali) [inline pkg/volume/util/goroutinemap/goroutinemap.go]: Caller is expected to acquire lock (added comment clarifying).

File: pkg/volume/util/goroutinemap/goroutinemap.go (hunk 24 above)
Comment (saad-ali) [inline pkg/volume/util/goroutinemap/goroutinemap.go]: Caller is expected to acquire lock (added comment clarifying).

File: pkg/util/goroutinemap/common/util.go (hunk 22 above)
Comment (saad-ali) [inline pkg/util/goroutinemap/common/util.go]: It sets the error so that a common error handling routine can be used to handle both errors and panics. I moved it to `pkg/util/runtime` and deleted this file.

File: pkg/volume/util/goroutinemap/goroutinemap.go (hunk 9 above)
Comment (saad-ali) [inline pkg/volume/util/goroutinemap/goroutinemap.go]: Spoke with @matchstick offline we settled on `nestedpendingoperations.go`

Comment (smarterclayton) [inline pkg/volume/util/goroutinemap/goroutinemap.go]: Sgtm

Comment (saad-ali) [inline pkg/volume/util/goroutinemap/goroutinemap.go]: Done. Not sure what you mean by "move it out of goroutinemap directory"--this file is already in `pkg/volume/util`

File: pkg/util/goroutinemap/common/errors.go (hunk 7 above)
Comment (saad-ali) [inline pkg/util/goroutinemap/common/errors.go]: Removed.

File: pkg/util/goroutinemap/goroutinemap.go (hunk 11 above)
Comment (saad-ali) [inline pkg/util/goroutinemap/goroutinemap.go]: Ack, reverted rename.

File: pkg/util/goroutinemap/common/exponential_backoff.go (hunk 12 above)
Comment (saad-ali) [inline pkg/util/goroutinemap/common/exponential_backoff.go]: Done.

File: pkg/controller/volume/attachdetach/reconciler/reconciler.go (hunk 16 above)
Comment (saad-ali) [inline pkg/controller/volume/attachdetach/reconciler/reconciler.go]: Moved.

File: pkg/controller/volume/attachdetach/reconciler/reconciler.go (hunk 17 above)
Comment (saad-ali) [inline pkg/controller/volume/attachdetach/reconciler/reconciler.go]: Done.

File: pkg/controller/volume/persistentvolume/controller.go (hunk 18 above)
Comment (saad-ali) [inline pkg/controller/volume/persistentvolume/controller.go]: Reverted rename.

File: pkg/util/goroutinemap/common/errors.go (hunk 19 above)
Comment (saad-ali) [inline pkg/util/goroutinemap/common/errors.go]: Done.

File: pkg/util/goroutinemap/common/errors.go (hunk 20 above)
Comment (saad-ali) [inline pkg/util/goroutinemap/common/errors.go]: Done.

File: pkg/volume/util/nestedpendingoperations/nestedpendingoperations.go [hunk 25]
```diff
@@ -0,0 +1,288 @@
...
+		// Operation completed without error, or exponentialBackOffOnError disabled
+		grm.deleteOperation(volumeName, podName)
+		if *err != nil {
+			// Log error
+			logOperationName := getOperationName(volumeName, podName)
+			glog.Errorf("operation %s failed with: %v",
+				logOperationName,
+				*err)
+		}
+	} else {
```
Comment (matchstick) [inline pkg/volume/util/nestedpendingoperations/nestedpendingoperations.go]: @saad-ali @smarterclayton 

return instead of else.

File: pkg/volume/util/nestedpendingoperations/nestedpendingoperations.go [hunk 26]
```diff
@@ -0,0 +1,288 @@
...
+			return NewAlreadyExistsError(operationName)
+		}
+
+		operationName := getOperationName(volumeName, podName)
+		if err := previousOp.expBackoff.SafeToRetry(operationName); err != nil {
+			return err
+		}
+	}
+
+	if opExists {
```
Comment (matchstick) [inline pkg/volume/util/nestedpendingoperations/nestedpendingoperations.go]: We can merge this if clause with the one above.

**Comment by matchstick** [comment]
@saad-ali @smarterclayton I only had two nits. Once addressed please label as LGTM

File: pkg/volume/util/nestedpendingoperations/nestedpendingoperations.go (hunk 25 above)
Comment (saad-ali) [inline pkg/volume/util/nestedpendingoperations/nestedpendingoperations.go]: Done.

File: pkg/volume/util/nestedpendingoperations/nestedpendingoperations.go (hunk 26 above)
Comment (saad-ali) [inline pkg/volume/util/nestedpendingoperations/nestedpendingoperations.go]: Done.

**Comment by smarterclayton** [comment]
No other comments

File: pkg/util/goroutinemap/exponentialbackoff/exponential_backoff.go [hunk 27]
```diff
@@ -0,0 +1,120 @@
...
+import (
+	"fmt"
+	"time"
+)
+
+const (
+	// initialDurationBeforeRetry is the amount of time after an error occurs
+	// that GoroutineMap will refuse to allow another operation to start with
+	// the same target (if exponentialBackOffOnError is enabled). Each
+	// successive error results in a wait 2x times the previous.
```
Comment (jingxu97) [inline pkg/util/goroutinemap/exponentialbackoff/exponential_backoff.go]: I think there might be performance issue with the current exponential backoff design. Consider a scenario to creat a rc consisting of N pods which all reference the same disk . After desired world populator adds the volume and pod information, reconciler starts issuing waitforattach/mount requests periodically.  In the first round, reconciler scans all the pod volumes and issues total N requests almost at the same time. With gorountinemap, only one of them can succeed and some others will have ExponentialBackoffError which cause them to wait the same amount of time (double the current loop duration) to retry. After passing the durationBeforeRetry reconciler will issue N or N-1 (if the first finishes) requests again at around the same time, causing them to double the same wait time again. The same step repeat in the following rounds, each time only one request can be granted, and the waiting time to retry keeps doubling until reaching the maxDurationBeforeRetry. From my initial experiments, looks like mount operation finishes very fast which reduce the above problem. But wait for attach operation seems takes longer time. One possible solution is to introduce randomness into the algorithm.

Comment (saad-ali) [inline pkg/util/goroutinemap/exponentialbackoff/exponential_backoff.go]: > With gorountinemap, only one of them can succeed 

With this change all of them will be able to run mount in parallel.

> some others will have ExponentialBackoffError

ExponentialBackoff does not apply when an operation is rejected because there is an existing operation (only when there is an execution error).

Comment (jingxu97) [inline pkg/util/goroutinemap/exponentialbackoff/exponential_backoff.go]: Here I am more focusing on the operations for attachable volumes. As you explained, if ExponentialBackoff only apply for execution error, it may not cause serious performance issue. But still, it may end up spending more time than necessary. For example, if the volume becomes attachable in 1 min, depending on the timing, it is possible for all N pods waiting for more than 3 mins before starting mount operations. Also it may have higher chances of operation is rejected because there is an existing operation. What will happen if there is a hanging operation in goroutine?

For volumes that are not attachable, since they can run in parallel without problem. Why need to run them through goroutine?

Comment (matchstick) [inline pkg/util/goroutinemap/exponentialbackoff/exponential_backoff.go]: This feels like a 1.4 conversation if the exponential behaviour has not
changed. I am not saying JinXu's idea don't have merit, but we do need the
fix in 1.3 and if this is how it worked in 1.2 let's decouple this
conversation.

mrubin

Comment (saad-ali) [inline pkg/util/goroutinemap/exponentialbackoff/exponential_backoff.go]: > if the volume becomes attachable in 1 min, depending on the timing, it is possible for all N pods waiting for more than 3 mins before starting mount operations

That's not correct. If multiple pods on a node are referencing the same attachable volume, the attach and mount device steps are serialized and only happen *once* for all the pods. Once a device is attached and globally mounted, subsequent pods only need to do the bind mount (which, with this change is parallelized). 

> Also it may have higher chances of operation is rejected because there is an existing operation.

That makes little difference in execution speed. The reconciler retries very quickly.

> What will happen if there is a hanging operation in goroutine

If an attachable volume fails to attach or complete global mount, regardless of if the operations fail or hang, they will block bind mounting. That is necessary and by design, an attachable volume can not be bind mounted until these operations succeed.

> For volumes that are not attachable, since they can run in parallel without problem. Why need to run them through goroutine?

They must be run in a goroutine so that the caller is not blocked. They are run using `nestedpendingoperations` to prevent the multiple operations from being triggered on the same pod/volume.

**Comment by saad-ali** [comment]
@k8s-bot test this issue: #22655

**Comment by ncdc** [comment]
@saad-ali can you backport this to 1.3?

**Comment by matchstick** [comment]
@ncdc @kubernetes/sig-storage The plan is to backport this to 1.3 and release it in 1.3.4 in a week. Before merging into the 1.3 release branch we want to let it soak in master to undergo some exposure in e2e testing and such. Please also feel free to test this. More coverage the better. Hope that is ok

**Comment by smarterclayton** [comment]
Since we'd be soaking against 1.3 we just wanted to avoid soaking a
different backport than what would end up in Kube.
